### PR TITLE
Add support for asynchronous generation

### LINKED ISSUE #16: Asynchronous interface
First of all, your package is great, thank you!

My issue is a question/suggestion: The official Node.JS documentation says that the crypto.randomBytes method can block the execution:

> The crypto.randomBytes() method will not complete until there is sufficient entropy available. This should normally never take longer than a few milliseconds. The only time when generating the random bytes may conceivably block for a longer period of time is right after boot, when the whole system is still low on entropy.

You can find a question whether should we use sync or async:
https://github.com/nodejs/help/issues/457

I don't have any performance issue yet, but I would like to use this package in production. In my use case I expect to have a lot of call to this function so it would be better if it did not block the execution. I know that a few milliseconds as the documentation says is not a lot for a single call but in a server environment a lot of few milliseconds call can block other tasks.

I suggest that we should have an interface (eg. cryptoRandomStringAsync) which does similar things but returns a Promise. What do you think?

If you like the idea but don't have time for this I can implement it and send a PR.

Thanks

### PR DESCRIPTION
**Pull request by callmehiphop**
Closes #16 

Figured I'd have a go at this! Apologies in advance if you hate it 🐱

### CODE REVIEW

**Comment by sindresorhus** [comment]
The code changes look good. Can you also document it in the readme?

**Comment by callmehiphop** [comment]
@sindresorhus absolutely! I forgot to add unit tests as well, have anything special in mind there? Otherwise I can just copy/pasta the current tests and add the `async`/`await` stuff.

**Comment by sindresorhus** [comment]
Yeah, just copy-paste, but you don’t need all as the main logic is already tested by the sync methods.

**Comment by callmehiphop** [comment]
I ended up catching a bug with my code when I added the tests, so I went ahead and make a small refactor in addition to the tests, but I think we should be good!

File: index.js [hunk 1]
```diff
@@ -1,9 +1,21 @@
 'use strict';
 const crypto = require('crypto');
+const pify = require('pify');
+
+const randomBytesAsync = pify(crypto.randomBytes);
```
Comment (sindresorhus) [inline index.js]: You can just use `util.promisify`.

File: index.js [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
 
 const urlSafeCharacters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~'.split('');
 const numericCharacters = '0123456789'.split('');
 
+const generateRandomAsync = async (byteLength, type, length) => {
+	const buf = await randomBytesAsync(byteLength);
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	const buffer = await randomBytesAsync(byteLength);
```

File: index.d.ts [hunk 3]
```diff
@@ -55,19 +55,36 @@ declare namespace cryptoRandomString {
...
-*/
-declare function cryptoRandomString(options?: cryptoRandomString.Options): string;
+	/**
+	Asynchronously generate a [cryptographically strong](https://en.wikipedia.org/wiki/Strong_cryptography) random string.
+
+	@returns A promise that resolves with a randomized string.
+
+	@example
+	```
+	import {async as cryptoRandomString} = require('crypto-random-string');
```
Comment (sindresorhus) [inline index.d.ts]: I don't think this is correct.

```suggestion
	import cryptoRandomString = require('crypto-random-string');
```

And then use it as `cryptoRandomString.async()`.

**Comment by sindresorhus** [comment]
@callmehiphop Friendly bump :)

**Comment by sindresorhus** [comment]
Closing for lack of activity.

**Pull request closed without merging**